README.md
# eos-browser-tools

# Description

This package currently contains two main components

  * eos-browser-tools: Endless-specific and browser-related tools
  * eos-google-chrome-helper: Wrapper application to run Google Chrome

## eos-browser-tools

This currently provides a handler for `webapp:<WM_CLASS>@<URI>` URIs,
which allows us to easily run chromium in application mode by
specifying the desired `WM_CLASS` and final address to load in the URI.

As this handler gets installed in the OS, it can handle URLs loaded
from unsandboxed environments (e.g. Facebook, WhatsApp) as well as
those from sandboxed flatpak applications, that rely on the Flatpak's
OpenURI portal.

If Google Chrome has been installed via the App Center (which requires
having the eos-google-chrome-helper package installed), this script
will consider using it instead of Chromium if it's set as the default
browser, otherwise Chromium will be used.

## eos-google-chrome-helper

This package provides a system level wrapper application to allow easily
downloading, installing and running Google Chrome on Endless OS.

This wrapper application does mainly two things when you click on the desktop icon:

  * Integrates with the App Center (GNOME Software) so that it gets open on the page
    for Google Chrome when you click on the desktop icon and hasn't been installed yet.

  * If Google Chrome has been previously installed (using flatpak as a delivery mechanism),
    the wrapper script launches chromium with its own sandbox (outside of flatpak), by
    calling a launcher script that is shipped along with the "headless" flatpak app.

This package provides the following elements:
  * `eos-google-chrome`: wrapper to either launch Chrome or the App Center.
  * `eos-google-chrome.png`: icon to integrate with the desktop.
  * `google-chrome.desktop`: application information according to the Desktop Entry
  Specification, to integrate with the shell. Note that we can't name it like the
  icon (i.e. eos-google-chrome.desktop) since that way Google Chrome would not be
  able to recognize itself when running as the default browser, which would end up
  with Chromium asking to set itself as the default each time it was run.

All this files will be installed, exceptionally, as part of the OSTree, so that the
icon and the wrapper app are available on the desktop at any time, either to run
the browser or to install it if not yet available.

## License

eos-browser-tools is Copyright (C) 2016, 2017 Endless Mobile, Inc.
and is licensed under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2 of
the License, or (at your option) any later version.

See the COPYING file for the full version of the GNU GPLv2 license


debian/changelog
eos-browser-tools (1.1.0-0endless1) eos; urgency=medium

  * Fix use of renamed malcontent-client API and bump the malcontent-tools
    dependency (T29195)

 -- Philip Withnall <withnall@endlessm.com>  Thu, 23 Jan 2020 13:48:00 +0000

eos-browser-tools (1.1.0) eos; urgency=medium

  * Bump version number so that we can vanquish the short-lived reign of
    eos-application-tools.

 -- Robert McQueen <rob@endlessm.com>  Fri, 04 Aug 2017 09:54:39 +0100

eos-browser-tools (1.0.1) eos; urgency=medium

  * Renamed to eos-browser-tools, so that we can ship browser related
    tools separately from the eos-google-chrome-helper package.

  https://phabricator.endlessm.com/T15917

 -- Mario Sanchez Prada <mario@endlessm.com>  Tue, 07 Mar 2017 11:42:24 +0000

eos-google-chrome-helper (1.0.0) eos; urgency=medium

  * Initial release

 -- Mario Sanchez Prada <mario@endlessm.com>  Thu, 10 Nov 2016 13:20:12 +0000


chrome-helper/eos-google-chrome-installer.py
#!/usr/bin/python3
#
# eos-google-chrome-installer: helper script to install Google Chrome
#
# Copyright (C) 2016, 2017 Endless Mobile, Inc.
# Authors:
#  Michal Rostecki <michal@kinvolk.io>
#  Mario Sanchez Prada <mario@endlessm.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import argparse
import config
import configparser
import logging
import os
import subprocess
import sys

import gi
gi.require_version('Flatpak', '1.0')
from gi.repository import Flatpak
from gi.repository import Gio
from gi.repository import GLib
from systemd import journal


def exit_with_error(*args):
    logging.error(*args)
    sys.exit(1)


class GoogleChromeInstaller:
    def __init__(self, initial_setup):
        self._initial_setup = initial_setup

        if self._initial_setup:
            if not self._automatic_install_enabled():
                logging.info("Google Chrome installation is disabled")
                sys.exit(0)

            if self._initial_setup_already_done():
                logging.info("Google Chrome automatic installation already done")
                sys.exit(0)

        try:
            self._installation = Flatpak.Installation.new_system()
        except GLib.Error as e:
            exit_with_error("Couldn't not find current system installation: %r", e)

        if self._check_chrome_flatpak_launcher():
            logging.info("Google Chrome is already installed")
            self._touch_done_file()
            return

        logging.info("Could not find flatpak launcher for Chrome.")

        if self._initial_setup:
            self._wait_for_network_connectivity()

        self._run_app_center_for_chrome()

    def _initial_setup_already_done(self):
        if os.path.exists(config.STAMP_FILE_INITIAL_SETUP_DONE):
            return True

        legacy_stamp_file = os.path.expanduser(config.LEGACY_USER_CONFIG_STAMP_FILE)
        if os.path.exists(legacy_stamp_file):
            # This is a legacy scenario, so we touch the system-wide file to
            # make sure next time that's the only thing that gets checked.
            self._touch_done_file()
            return True

        return False

    def _automatic_install_enabled(self):
        if not os.path.exists(config.CONFIG_FILE):
            logging.warning("Could not find configuration file at {}"
                            .format(config.CONFIG_FILE))
            return False

        is_enabled = False
        with open(config.CONFIG_FILE, 'r') as config_file:
            helper_config = configparser.ConfigParser(allow_no_value=True)
            try:
                helper_config.read_file(config_file)
                logging.info("Read contents from configuration file at {}\n"
                             .format(config.CONFIG_FILE))
            except configparser.ParsingError as e:
                logging.error("Error parsing contents from configuration file at {}: {}"
                             .format(config.CONFIG_FILE, str(e)))
                return False

            try:
                is_enabled = helper_config.getboolean('Initial Setup', 'AutomaticInstallEnabled')
                logging.info("AutomaticInstallEnabled = {}".format(str(is_enabled)))
            except configparser.NoOptionError:
                logging.warning("AutomaticInstallEnabled key not found in {}".format(config.CONFIG_FILE))
                return False

        return is_enabled

    def _check_chrome_flatpak_launcher(self):
        try:
            self._installation.get_current_installed_app(config.FLATPAK_CHROME_APP_ID, None)
        except GLib.Error as e:
            logging.info("Chrome application is not installed")
            return False
        return True

    def _is_connected_state(self):
        monitor = Gio.NetworkMonitor.get_default()
        return monitor.get_connectivity() == Gio.NetworkConnectivity.FULL

    def _wait_for_network_connectivity(self):
        def _network_changed(monitor, available, loop):
            if not available:
                logging.info("No network available")
                return

            if monitor.get_connectivity() != Gio.NetworkConnectivity.FULL:
                logging.info("Network available, but not connected to the Internet")
                return

            logging.info("Connected to the network and the internet")
            loop.quit()

        logging.info("Checking network connectivity...")
        if self._is_connected_state():
            logging.info("Network connected")
            return

        logging.info("Not connected to any network, wait for connection")

        loop = GLib.MainLoop()

        monitor = Gio.NetworkMonitor.get_default()
        monitor.connect('network-changed', _network_changed, loop)
        loop.run()

    def _wait_for_installation(self):
        def _installation_finished(monitor, file_, other_file, event_type):
            if event_type != Gio.FileMonitorEvent.CHANGES_DONE_HINT:
                return

            if self._check_chrome_flatpak_launcher():
                logging.info("{} has been installed".format(config.FLATPAK_CHROME_APP_ID))
                loop.quit()

        loop = GLib.MainLoop()

        monitor = self._installation.create_monitor(None)
        monitor.connect('changed', _installation_finished)

        loop.run()

    def _set_as_default_browser(self):
        mimetypes = ['text/html',
                     'x-scheme-handler/http',
                     'x-scheme-handler/https',
                     'x-scheme-handler/about']
        for mimetype in mimetypes:
            try:
                subprocess.check_call('xdg-mime default google-chrome.desktop {}'.format(mimetype),
                                      shell=True)
            except subprocess.CalledProcessError as e:
                exit_with_error("Couldn't start xdg-mime: {}".format(str(e)))
        try:
            subprocess.check_call('xdg-settings set default-web-browser google-chrome.desktop',
                                  shell=True)
        except subprocess.CalledProcessError as e:
            exit_with_error("Couldn't start xdg-settings: {}".format(str(e)))

    def _touch_done_file(self):
        # The system-wide stamp file touched by this helper makes sure that
        # the automatic installation won't ever be performed for other users.
        system_helper_cmd = os.path.join(config.PKG_DATADIR, 'eos-google-chrome-system-helper.py')
        try:
            subprocess.check_call('pkexec {}'.format(system_helper_cmd), shell=True)
        except subprocess.CalledProcessError as e:
            exit_with_error("Couldn't run {}: {}".format(system_helper_cmd, str(e)))

    def _post_install_chrome(self):
        self._wait_for_installation()

        if not self._check_chrome_flatpak_launcher():
            exit_with_error("Chrome isn't installed - something went wrong in GNOME Software")

        logging.info("Chrome successfully installed")

        self._set_as_default_browser()
        self._touch_done_file()

        logging.info("Post-installation configuration done")


    def _get_unique_id(self):
        chrome_app_center_id = config.FLATPAK_CHROME_APP_ID

        default_branch = None
        try:
            remote = self._installation.get_remote_by_name(config.FLATPAK_REMOTE_EOS_APPS)
        except GLib.Error as e:
            logging.warning("Could not find flatpak remote {}: {}"
                            .format(config.FLATPAK_REMOTE_EOS_APPS, str(e)))

        # Get the default branch now to construct the full unique ID GNOME Software expects.
        default_branch = remote.get_default_branch()
        if default_branch:
            chrome_app_center_id = 'system/flatpak/{}/desktop/{}/{}'.format(config.FLATPAK_REMOTE_EOS_APPS,
                                                                            config.FLATPAK_CHROME_APP_ID,
                                                                            default_branch)
        return chrome_app_center_id

    def _run_app_center_for_chrome(self):
        # FIXME: Ideally, we should be able to pass 'com.google.Chrome' to GNOME Software
        # and it would do the right thing by opening the page for the app's branch matching
        # the default branch for the apps' source remote. Unfortunately, this is not the case
        # at the moment and fixing it is non-trivial, so we'll construct the full unique ID
        # that GNOME Software expects, right from here, based on the remote's metadata.
        unique_id = self._get_unique_id()

        logging.info("Opening App Center...")
        if self._initial_setup:
            app_center_argv = ['gnome-software', '--install', unique_id, '--interaction', 'none']
        else:
            app_center_argv = ['gnome-software', '--details={}'.format(unique_id)]

        try:
            subprocess.Popen(app_center_argv)
        except OSError as e:
            exit_with_error("Could not launch Chrome: {}".format(repr(e)))

        # There's a post-install procedure for automatic installations.
        if self._initial_setup:
            self._post_install_chrome()


def main():
    # Send logging messages both to the console and the journal
    logging.basicConfig(level=logging.INFO)
    logging.root.addHandler(journal.JournalHandler())

    parser = argparse.ArgumentParser()
    parser.add_argument('--debug', dest='debug', action='store_true')
    parser.add_argument('--initial-setup', dest='initial_setup', action='store_true')

    parsed_args = parser.parse_args()

    if parsed_args.debug:
        logging.root.setLevel(logging.DEBUG)

    app_arch = Flatpak.get_default_arch()
    if app_arch != 'x86_64':
        exit_with_error("Found installation of unsupported architecture: %s", app_arch)

    GoogleChromeInstaller(parsed_args.initial_setup)
    sys.exit(0)


if __name__ == '__main__':
    main()


chrome-helper/eos-google-chrome-system-helper.py
#!/usr/bin/python3
#
# eos-google-chrome-system-helper: system helper to touch / reset the stamp file
#
# Copyright (C) 2017 Endless Mobile, Inc.
# Authors:
#  Mario Sanchez Prada <mario@endlessm.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import argparse
import config
import logging
import os
import sys

from systemd import journal


def exit_with_error(*args):
    logging.error(*args)
    sys.exit(1)


def touch_file(file):
    with open(file, 'a'):
        os.utime(file, None)


def create_stamp_file(filename):
    dirname = os.path.dirname(filename)

    try:
        os.makedirs(dirname, exist_ok=True)
        touch_file(filename)
        logging.info("Stamp file created at {}".format(filename))
    except OSError as e:
        exit_with_error("Error creating stamp file at {}: {}"
                        .format(filename, str(e)))


def remove_stamp_file(filename):
    try:
        os.unlink(filename)
        logging.info("Stamp file removed from {}".format(filename))
    except OSError as e:
        exit_with_error("Error removing stamp file at {}: {}"
                        .format(filename, str(e)))


if __name__ == '__main__':
    # Send logging messages both to the console and the journal
    logging.basicConfig(level=logging.INFO)
    logging.root.addHandler(journal.JournalHandler())

    parser = argparse.ArgumentParser(description="Marks Google Chrome installation as installed, system-wide")

    parser.add_argument('--debug', help="Show extra messages", action='store_true')
    parser.add_argument('--reset', help="Remove stamp file", action='store_true')

    parsed_args = parser.parse_args()
    if parsed_args.debug:
        logging.root.setLevel(logging.DEBUG)

    if parsed_args.reset:
        remove_stamp_file(config.STAMP_FILE_INITIAL_SETUP_DONE)
    else:
        create_stamp_file(config.STAMP_FILE_INITIAL_SETUP_DONE)

    sys.exit(0)


chrome-helper/eos-google-chrome.py
#!/usr/bin/python3
#
# eos-google-chrome: helper script to install/launch Google Chrome
#
# Copyright (C) 2016, 2017 Endless Mobile, Inc.
# Authors:
#  Mario Sanchez Prada <mario@endlessm.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import argparse
import logging
import os
import subprocess
import sys

import config
import gi
gi.require_version('Flatpak', '1.0')
from gi.repository import Flatpak
from gi.repository import GLib
from systemd import journal


def exit_with_error(message):
    logging.error(message)
    sys.exit(1)


class GoogleChromeLauncher:
    def __init__(self, params):
        self._params = params
        try:
            self._installation = Flatpak.Installation.new_system()
        except GLib.Error as e:
            exit_with_error("Could not find current system installation: {}".format(repr(e)))

        self._start()

    def _start(self):
        chrome_launcher = self._get_chrome_flatpak_launcher()
        if chrome_launcher:
            logging.info("Flatpak launcher for Chrome found. Launching...")
            self._run_chrome_app(chrome_launcher, self._params)
        else:
            logging.info("Could not find flatpak launcher for Chrome. Running installation script...")
            self._install_chrome()

    def _run_chrome_app(self, chrome_launcher, params):
            try:
                launcher_process = subprocess.Popen([chrome_launcher] + params)
                logging.info("Running Google Chrome launcher with PID %d", launcher_process.pid)
            except OSError as e:
                exit_with_error("Could not launch Google Chrome: {}".format(repr(e)))

    def _install_chrome(self):
        try:
            subprocess.Popen([os.path.join(config.PKG_DATADIR, 'eos-google-chrome-installer.py')])
        except OSError as e:
            exit_with_error("Could not launch Chrome: {}".format(repr(e)))

    def _get_chrome_flatpak_launcher(self):
        app = None
        try:
            app = self._installation.get_current_installed_app(config.FLATPAK_CHROME_APP_ID, None)
        except GLib.Error:
            logging.info("Chrome application is not installed")
            return None

        app_path = app.get_deploy_dir()
        if not app_path or not os.path.exists(app_path):
            exit_with_error("Could not find Chrome's application directory")

        app_launcher_path = os.path.join(app_path, 'files', 'bin', 'eos-google-chrome-app')
        if not os.path.exists(app_launcher_path):
            exit_with_error("Could not find flatpak launcher for Google Chrome")

        logging.info("Found flatpak launcher for Google Chrome: %s", repr(app_launcher_path))
        return app_launcher_path


if __name__ == '__main__':
    # Send logging messages both to the console and the journal
    logging.basicConfig(level=logging.INFO)
    logging.root.addHandler(journal.JournalHandler())

    parser = argparse.ArgumentParser()
    parser.add_argument('--debug', dest='debug', action='store_true')

    parsed_args, otherargs = parser.parse_known_args()

    if parsed_args.debug:
        logging.root.setLevel(logging.DEBUG)

    # Google Chrome is only available for Intel 64-bit
    app_arch = Flatpak.get_default_arch()
    if app_arch != 'x86_64':
        exit_with_error("Found installation of unsupported architecture: {}".format(app_arch))


    GoogleChromeLauncher(otherargs)
    sys.exit(0)
